Unit Tests Framework

Ccomp5950 opened the pull request:
Adds an example, edits Travis to run them.

Travis will now start up the server then start some tests on a live server when a round started in extended.

Further Reading: https://en.wikipedia.org/wiki/Unit_testing

Ccomp5950 committed:
Unit Tests

Adds an example, edits Travis to run them.

GinjaNinja32 commented on .travis.yml:
```diff
@@ -40,5 +40,8 @@ script:
   - python tools/TagMatcher/tag-matcher.py ../..
   - python tools/GenerateChangelog/ss13_genchangelog.py html/changelog.html html/changelogs
   - source $HOME/BYOND-${BYOND_MAJOR}.${BYOND_MINOR}/byond/bin/byondsetup
+  - echo "#define UNIT_TEST 1" >> code/__defines/unit_testing.dm
```
`>>` should likely just be a single `>`, unless you want two defines of it (which will work because of how BYOND works, but seems unintended).

Ccomp5950 replied:
Good catch, supposed to be a single > (overwrite entire file) instead of >> (append to end of file)

Ccomp5950 committed:
Fixed travis.yml to have the correct redirection

Ccomp5950 committed:
Unit Test: Escape Shuttle ZAS check. + Templates

Ccomp5950 committed:
Adds Supply Shuttle Unit Tests.

Moves the Supply Shuttle and tests air in both locations.

turned the air test into a "helper" proc.

Changed the world.dm message when unit testing is enabled so folks aren't caught off guard when the server shutsdown after testing.

GinjaNinja32 commented on code/unit_tests/zas_tests.dm:
```diff
@@ -0,0 +1,125 @@
+        name = "ZAS: Escape Shuttle"
+
+datum/unit_test/zas_escape_shuttle/start_test()
+	var/list/result = test_air_in_area("/area/shuttle/escape/centcom")
```
Passing strings then `text2path()`-ing them? Why not just pass paths to start with?

Ccomp5950 committed:
Removed text2path.

Kudos PsiOmegaDelta

Ccomp5950 committed:
Updated ZAS unit test to be more re-useable.

Went nuts adding some.  This will fail travis due to some existing bugs.

Ccomp5950 committed:
Vacuum not Vacume.

PsiOmegaDelta commented:
I assume the failing test is supposed to pass if you merge in the asteroid fix from dev.

Ccomp5950 commented:
No, I'm not sure why the mining turf is failing, it's something I went "welp will need to look at it tomorrow."

Ccomp5950 committed:
Made failed unit tests a little more noticeable.

Ccomp5950 committed:
Unit Tests now sets some config options if they aren't already set.

Ccomp5950 committed:
Terminal Colors Green and Red for success states.

Ccomp5950 committed:
Change to _template.

Ccomp5950 committed:
Adding mob damage tests.  -- This will fail in travis --

Species damage modifications haven't been completed in here.  And IPC's are dumb.

Ccomp5950 committed:
Finalized mob damage procs.

I'm done.

Ccomp5950 commented:
This is done.  I'll be adding more unit tests later and I hope others do as well but for now this is ready to be merged after reviews.

PsiOmegaDelta commented:
:+1:

HiddenKn commented on code/unit_tests/mob_tests.dm:
```diff
@@ -0,0 +1,487 @@
+	if(starting_oxyloss < ending_oxyloss)
+		pass("Oxyloss = [ending_oxyloss]")
+	else
+		fail("Mob is not taking oxygen damage.  Damange is [ending_oxyloss]")
```
Typo? Damange -> Damage

Kasuobes replied:
You dug up a one year old PR just for this...?

I mean, such dedication.

Ccomp5950 replied:
I no type guud
